### PR TITLE
Add Cube Utilization statistic to Better Victory Screen

### PR DESCRIPTION
**Pull request by heinwessels**
I think I found a nice way to calculate the cube utilization percentage. The only disadvantage is that it's not very efficient when it's carried by a player. But seeing can only be carried by one player then it shouldn't be a problem. Also, it's still plenty fast. If you're nervous about the performance best to do the check less often, like once every 20 ticks, but I think it's good enough.

And the statistic looks cool too
image

### CODE REVIEW

File: scripts/cube_fx.lua [hunk 1]
```diff
@@ -290,11 +300,59 @@ local function cube_victory(size, results)
...
+-- the cube. 
+local cube_working_machine_types = {
+  ["character"]           = cube_machine_condition_handlers["handcrafting"],
+  ["assembling-machine"]  = cube_machine_condition_handlers["status"],
+  ["furnace"]             = cube_machine_condition_handlers["status"],
+  ["boiler"]              = cube_machine_condition_handlers["status"],
+  ["generator"]           = cube_machine_condition_handlers["status"],
+  ["burner-generator"]    = cube_machine_condition_handlers["status"],
+  ["rocket-silo"]         = cube_machine_condition_handlers["status"],
+  ["reactor"]             = cube_machine_condition_handlers["status"], -- Does this work?
```
Comment (grandseiken) [inline scripts/cube_fx.lua]: I think:
* for machines that can burn the cube it'd be better to check if `entity.burner.currently_burning` is the item in question (otherwise the cube sitting in the fuel input/output slots counts as being used).
* for machines that use the cube as an ingredient it'd be better to check if it's working and the current recipe is a cube recipe (this is easy to check by seeing if the recipe is in the `cube_management.recipes()` map). This fixes some rare cases (e.g. cube sitting in the output slot of a fuel refinery crafting a non-cube recipe).

File: scripts/cube_fx.lua [hunk 2]
```diff
@@ -310,10 +368,20 @@ local function track_victory_statistics(size, results)
     end
   end
   victory_statistics.cube_last_position = new
+
+  -- Track cube "utilization"
+  local machine_conditional = cube_working_machine_types[entity.type]
+  if machine_conditional and machine_conditional(entity, results[1].item) then
+    -- The cube is in some entity that's using the cube "utilization"
+    victory_statistics.utilization.working = victory_statistics.utilization.working + update_tick_modulus
```
Comment (grandseiken) [inline scripts/cube_fx.lua]: You could just add `1` instead of `update_tick_modulus` and get the same result, then we don't need that variable.

File: scripts/cube_fx.lua [hunk 3]
```diff
@@ -310,10 +368,20 @@ local function track_victory_statistics(size, results)
     end
   end
   victory_statistics.cube_last_position = new
+
+  -- Track cube "utilization"
+  local machine_conditional = cube_working_machine_types[entity.type]
+  if machine_conditional and machine_conditional(entity, results[1].item) then
```
Comment (grandseiken) [inline scripts/cube_fx.lua]: If there's more than 1 cube we could randomly choose the index to check for utilization to get a less biased sampling.

Could also check them all and multiply the amounts we're adding to `working`/`idle` by the `search_result_weight` of the item defined in `cube_search.lua` but that's more expensive.

File: scripts/cube_fx.lua [hunk 4]
```diff
@@ -290,11 +300,59 @@ local function cube_victory(size, results)
...
+  -- We will only look at what the first cube is doing, because there might be
+  -- multiple like when it's in its phantom form. This makes it not entirely
+  -- accurate, but it's better than nothing. It's also important that we track
+  -- something as often as possible to ensure the "utilization" is calculated
+  -- somewhat correctly.
   local entity = results[1].entity
+
   if not entity or not entity.valid then return end
 
+  -- Track distance travelled by this cube
```
Comment (grandseiken) [inline scripts/cube_fx.lua]: I think we cannot attempt to track the distance travelled when there are multiple cubes since the ordering of results is not guaranteed to be consistent. The first result may be the position of some cube one tick and the position of a totally different cube the next. I think we should just clear the `cube_last_position` whenever there is more than one.

File: scripts/cube_fx.lua [hunk 5]
```diff
@@ -290,11 +300,59 @@ local function cube_victory(size, results)
   end
 end
 
+-- Map to determine if some machine is using the cube
+-- in an "efficient" manner. 
+local cube_machine_condition_handlers = {
```
Comment (grandseiken) [inline scripts/cube_fx.lua:334]: I'd prefer if the logic was just inlined into the `track_victory_statistics` function, like
```
if crafter_entity_types[entity_type] then
  // crafter utilization code
elseif burner_types[entity_type] then
  // burner utilization code
end
```
unless you have good reason to think your way is better for performance. I suspect function call overhead would be more than some map checks.

File: locale/en/better-victory-screen.cfg [hunk 6]
```diff
@@ -3,10 +3,14 @@ ultracube=Ultracube
 
 [bvs-stats]
 cube-distance-travelled=Distance travelled by cube
+cube-utilization=Cube Utilization
```
Comment (grandseiken) [inline locale/en/better-victory-screen.cfg]: nit: lowercase utilization for consistency with factorio naming schemes

File: locale/en/better-victory-screen.cfg [hunk 7]
```diff
@@ -3,10 +3,14 @@ ultracube=Ultracube
...
 cubes-consumed=Ultradense cubes consumed
 cubes-consumed-dormant=Dormant cubes consumed
 cubes-consumed-phantom=Phantom cubes consumed
 cubes-consumed-phantom-dormant=Dormant phantom cubes consumed
 cubes-reconstructed=Cubes destroyed and reconstructed
 cubes-consumed-total=Total cubes consumed
 matter-created=Matter created
+
+[bvs-stat-tooltip]
+cube-effiency=Percentage of time the cube was utilized not not simply idling.
```
Comment (grandseiken) [inline locale/en/better-victory-screen.cfg]: not not?

File: scripts/cube_fx.lua (hunk 5 above)
Comment (heinwessels) [reply scripts/cube_fx.lua:334]: I did it for performance reasons. If you use a large `if ellse` chain then every `if` needs to be executed until the one we need is found. Whereas with my approach if can find the function directly.

That said, this list likely will be less than 10 long, so I'm not sure if it's really that much worse performance-wise.

File: scripts/cube_fx.lua (hunk 2 above)
Comment (heinwessels) [reply scripts/cube_fx.lua]: Yeah true, realized that a few hours after making the PR. Changed to `1`.

File: scripts/cube_fx.lua (hunk 1 above)
Comment (heinwessels) [reply scripts/cube_fx.lua]: I don't think this is required. We know:
 - If the entity is in a `assembly-machine`, `furnace` or `generator` is _has_ to be either an ingredient or the fuel.
 - If this is the case the machine can only be actively running if the cube is is being used. And in this case the status will always be `working`. All other cases have different statusses:
   - If it's an ingredient and it's waiting for other ingredients it will be `item_ingredient_shortage` or `no_ingredients`.
   - If it's in the output of a crafter it will be `full_output`.
   - If it's the fuel (furnace or burner), but it cannot be burned yet do to ingredient shortage, or the output being full, then the status will be what's mentioned above.
   - When in the output slot of the nuclear reactor it says `no fuel`
   
> This fixes some rare cases (e.g. cube sitting in the output slot of a fuel refinery crafting a non-cube recipe).

This is possible. If you change it to a non-cube recipe then the cube will be taken out of the entity. If the cube is in a crafter, then the crafter has to have a cube recipe set.
   
Only thing I can think of is maybe while launching the rocket, but I'm no t sure if the silo is used.

I might be missing some other edge case? But to me it seems very accurate and simple

File: scripts/cube_fx.lua (hunk 4 above)
Comment (heinwessels) [reply scripts/cube_fx.lua]: Yeah, you're right. Good catch. Will stop tracking when the cube is split apart.

File: scripts/cube_fx.lua (hunk 3 above)
Comment (heinwessels) [reply scripts/cube_fx.lua]: I went for choosint a random cube to keep the processing to a minimum.

File: scripts/cube_fx.lua (hunk 1 above)
Comment (grandseiken) [reply scripts/cube_fx.lua]: > If the entity is in a assembly-machine, furnace or generator is has to be either an ingredient or the fuel.

This isn't true. The player can (manually) place the cube wherever they want. They can put the cube in an output slot of a fabricator that happens to be crafting something unrelated. There's no reason to do that and it's weird, but right now the code would count that as the cube being used.

Comment (heinwessels) [reply scripts/cube_fx.lua]: I'll have a look tonight if I can get it into a state like this :)

Comment (heinwessels) [reply scripts/cube_fx.lua]: > They can put the cube in an output slot of a fabricator that happens to be crafting something unrelated

I still don't think this is possible :) 

image
image

---

What's maybe possible is the following case where you put the cube in a furnace output slot. But when it's an `ultra-dense` then it's impossible that's it's currently burning, meaning status is always `No Fuel` as seen in this screen grab. One case where this isn't true is with the phantom cubes, by placing a phantom cube in the output slot, but that would mean that another phantom cube is burning, and this it's status would be `working` correctly.
image

---

I'm sorry if I'm coming across stubborn. I just feel that any other approach with specific tests for every entity type would introduce more potential bugs and edge cases, as well as being worse for performance. Whereas this approach is easy to understand, fast, and fool-proof.

That said, this is your code, and I will change it if you feel strongly about it :)

Comment (heinwessels) [reply scripts/cube_fx.lua]: Wait, one valid edge case is placing the cube in the boiler output slot, and then feeding the boiler with `Condensed Fuel`. I'll add a special case for the boiler.
image

File: scripts/cube_fx.lua [hunk 8]
```diff
@@ -304,26 +308,122 @@ local function cube_victory(size, results)
...
+  -- then it has to be an ingredient or fuel. And thus, if the cube is in a machine
+  -- and it's not currently "working" then it would mean the cube is idling due to
+  -- `no_fuel`, `output_full`, `no_ingredients`, etc.
+  -- One edge case might be phantom cube placed in the output slot of a boiler
+  -- while the boiler is burning, but the only possible fuel source then
+  ["assembling-machine"]  = cube_machine_condition_handlers["status"],
+  ["furnace"]             = cube_machine_condition_handlers["status"],
+  ["generator"]           = cube_machine_condition_handlers["status"],
+  ["burner-generator"]    = cube_machine_condition_handlers["status"],
+  ["rocket-silo"]         = cube_machine_condition_handlers["status"],
```
Comment (grandseiken) [inline scripts/cube_fx.lua]: The rocket silo is a teleporter here, and the cube can only be in it if it's waiting to be teleported. Not sure this should count as usage?

File: scripts/cube_fx.lua (hunk 1 above)
Comment (grandseiken) [reply scripts/cube_fx.lua]: Ok, you're right that you can't put the cube in assembly machine output slots, I was confusing their behaviour with burnt fuel result slots.

It's no problem if you're being stubborn since I am too :)

However, here's 2 more edge cases:
- Dormant cube sitting in the output slot of a working mystery furnace
- Any cube sitting in the burnt result slot of a nuclear reactor that's online and now working with uranium fuel cells

I'm also thinking of adding endgame technologies allowing alternative ways to split the cube which might introduce more edge cases. That's one reason I was arguing for checking things a little more thoroughly, so we rely less on assumptions about exactly how things are set up now. What do you think?

Comment (heinwessels) [reply scripts/cube_fx.lua]: Okay, added another category `crafter` that works for `assembling-machines`, `furnaces`, (and `silos`?). It checks if the cube is being burned, and then if it's an ingredient :)

File: scripts/cube_fx.lua (hunk 8 above)
Comment (heinwessels) [reply scripts/cube_fx.lua]: I'm not sure what the timing is like using the teleporter, but if there's a launching animation we could set it as `used` while the animation is ongoing? Can't play with it though cause it currently crashes :)

Comment (heinwessels) [reply scripts/cube_fx.lua]: Just played with it a little and IMO this should count as `idle`, because the cube is only being transported, not used The same reason IMO we're ignoring inserters, cars, trains, etc. So for now I'm removing the `rocket-silo` entry, so that it's always counted as idle.

**Comment by grandseiken** [comment]
Thanks for humouring me!
